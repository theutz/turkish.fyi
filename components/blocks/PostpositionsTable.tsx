import { HTMLProps, ReactNode, useCallback, useEffect, useState } from "react";
import { Postpositions, Translation } from "../../data/postpositions";

type Props = {
  data?: Postpositions;
};

function TH({
  isVisible = true,
  ...props
}: {
  isVisible?: boolean;
} & HTMLProps<HTMLTableHeaderCellElement>) {
  return isVisible ? <th {...props} /> : null;
}

export function useFieldVisibility() {
  const defaultFields = ["word", "root", "examples", "kind", "group"];
  const [fields, setFields] = useState(defaultFields);

  const isVisible = (field: typeof defaultFields[0]) => fields.includes(field);
  const hideField = (field: typeof defaultFields[0]) => {
    setFields((prev) => prev.filter((f) => f !== field));
  };

  return { isVisible, hideField };
}

export function PostpositionsTable({ data = [] }: Props) {
  const { isVisible, hideField } = useFieldVisibility();

  if (data.length === 0) return null;

  return (
    <table className="border-collapse table-auto">
      <thead>
        <tr className="text-left border-b border-gray-400">
          <TH isVisible={isVisible("word")} onClick={() => hideField("word")}>
            Word
          </TH>
          <TH isVisible={isVisible("word")}>Definition</TH>
          <TH isVisible={isVisible("kind")}>Kind</TH>
          <TH isVisible={isVisible("group")}>Group</TH>
          <TH isVisible={isVisible("root")} colSpan={2}>
            Root
          </TH>
          <TH isVisible={isVisible("examples")} colSpan={2}>
            Examples
          </TH>
        </tr>
      </thead>
      <tbody>
        {data.map(
          ({
            word: [word, definition = []],
            kind,
            group,
            root: [root, rootDef = []] = [],
            examples = [],
          }) => {
            const longest = getLongestArray(definition, examples, rootDef);
            const makeCellComponent = makeCellComponentFactory(longest);

            return Array.from({ length: longest }, (_, i) => i).map((i) => {
              const makeTestId = makeTestIdFactory(word, kind, group, i);
              const rowId = makeTestId()();
              const Cell = makeCellComponent({ i, makeTestId });

              return (
                <tr
                  key={rowId}
                  data-testid={rowId}
                  className="border-b border-b-gray-200"
                >
                  <Cell isVisible={isVisible("word")} fieldName="word">
                    {word}
                  </Cell>
                  <Cell isVisible={isVisible("word")} fieldName="definition">
                    {definition}
                  </Cell>
                  <Cell isVisible={isVisible("kind")} fieldName="kind">
                    {kind}
                  </Cell>
                  <Cell isVisible={isVisible("group")} fieldName="group">
                    {group}
                  </Cell>
                  <Cell isVisible={isVisible("root")} fieldName="root">
                    {root}
                  </Cell>
                  <Cell isVisible={isVisible("root")} fieldName="rootDef">
                    {rootDef}
                  </Cell>
                  <Cell isVisible={isVisible("examples")} fieldName="examples">
                    {examples}
                  </Cell>
                </tr>
              );
            });
          }
        )}
      </tbody>
    </table>
  );
}

function makeCellComponentFactory(longestEntryLength: number) {
  return ({
    i,
    makeTestId,
  }: {
    i: number;
    makeTestId: ReturnType<typeof makeTestIdFactory>;
  }) => {
    return function Cell({
      children,
      fieldName,
      isVisible = true,
    }: {
      children?: string | readonly string[] | readonly Translation[];
      fieldName: string;
      isVisible?: boolean;
    }) {
      if (!isVisible) {
        return null;
      }

      const id = makeTestId(fieldName)();

      if (["string", "undefined"].includes(typeof children)) {
        if (i === 0) {
          return (
            <td data-testid={id} rowSpan={longestEntryLength}>
              {children}
            </td>
          );
        }
      }

      if (Array.isArray(children)) {
        const rowSpan =
          i === children.length - 1
            ? longestEntryLength - children.length + 1
            : 1;

        if (i <= children.length - 1) {
          const child = children[i];

          if (Array.isArray(child)) {
            return (
              <>
                {child.map((c, j) => {
                  const content = Array.isArray(c) ? c[0] : c;

                  return (
                    <td
                      key={makeTestId(fieldName)(j)}
                      data-testid={makeTestId(fieldName)(j)}
                      rowSpan={rowSpan}
                    >
                      {content}
                    </td>
                  );
                })}
              </>
            );
          }

          return (
            <td data-testid={id} rowSpan={rowSpan}>
              {child}
            </td>
          );
        }

        if (children.length === 0) {
          return <td data-testid={id} rowSpan={1} />;
        }
      }

      return null;
    };
  };
}

type makeTestIdFactory = (
  word: string,
  kind: string,
  group: string,
  i: number
) => (field?: string) => (j?: number) => string;
export const makeTestIdFactory: makeTestIdFactory =
  (word, kind, group, i) => (field) => (j) =>
    [word, kind, group, i, field, j]
      .filter((x) => (x ?? null) !== null)
      .join("-");

function sortArrayByLengthDesc(a: readonly unknown[], b: readonly unknown[]) {
  if (a.length < b.length) return 1;
  if (a.length > b.length) return -1;
  return 0;
}

export function getLongestArray(...args: (readonly unknown[])[]): number {
  return args.sort(sortArrayByLengthDesc)[0].length;
}

export default PostpositionsTable;
