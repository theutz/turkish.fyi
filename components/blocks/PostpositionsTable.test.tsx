import { render } from "@testing-library/react";
import { PostpositionsTable as Sut } from "./PostpositionsTable";

test("works", () => {
  const { asFragment } = render(<Sut />);
  expect(asFragment()).toMatchSnapshot();
});
