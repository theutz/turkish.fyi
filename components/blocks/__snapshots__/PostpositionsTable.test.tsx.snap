// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
<DocumentFragment>
  <table
    class="border-collapse table-auto"
  >
    <thead>
      <tr
        class="border-b border-gray-400 text-left"
      >
        <th>
          Word
        </th>
        <th>
          Definition
        </th>
        <th>
          Kind
        </th>
        <th>
          Group
        </th>
        <th
          colspan="2"
        >
          Root
        </th>
        <th
          colspan="2"
        >
          Examples
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          gibi
        </td>
        <td
          rowspan="1"
        >
          like
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe gibi
        </td>
        <td
          rowspan="1"
        >
          like Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          için
        </td>
        <td
          rowspan="1"
        >
          for
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe için
        </td>
        <td
          rowspan="1"
        >
          for Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          ile
        </td>
        <td
          rowspan="1"
        >
          with
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe ile
        </td>
        <td
          rowspan="1"
        >
          with Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          kadar
        </td>
        <td
          rowspan="1"
        >
          as ... as
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe kadar ...
        </td>
        <td
          rowspan="1"
        >
          as ... as Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="2"
        >
          doğru
        </td>
        <td
          rowspan="2"
        >
          towards (time/place)
        </td>
        <td
          rowspan="2"
        >
          bare
        </td>
        <td
          rowspan="2"
        >
          dative
        </td>
        <td
          rowspan="2"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          üçe doğru
        </td>
        <td
          rowspan="1"
        >
          towards 3 o'clock
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          eve doğru
        </td>
        <td
          rowspan="1"
        >
          towards home
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="3"
        >
          göre
        </td>
        <td
          rowspan="1"
        >
          according to
        </td>
        <td
          rowspan="3"
        >
          bare
        </td>
        <td
          rowspan="3"
        >
          dative
        </td>
        <td
          rowspan="3"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          sana göre
        </td>
        <td
          rowspan="1"
        >
          according to you
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          compared with
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          İngeltere'ye göre
        </td>
        <td
          rowspan="1"
        >
          compared with England
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          right for
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          bize göre
        </td>
        <td
          rowspan="1"
        >
          right for us
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="2"
        >
          kadar
        </td>
        <td
          rowspan="1"
        >
          until, by (time)
        </td>
        <td
          rowspan="2"
        >
          bare
        </td>
        <td
          rowspan="2"
        >
          dative
        </td>
        <td
          rowspan="2"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          sabah kadar
        </td>
        <td
          rowspan="1"
        >
          until morning
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          as far as (place)
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ankara'ya kadar
        </td>
        <td
          rowspan="1"
        >
          as far as Ankara
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="3"
        >
          üstünde
        </td>
        <td
          rowspan="1"
        >
          on
        </td>
        <td
          rowspan="3"
        >
          posessive
        </td>
        <td
          rowspan="3"
        >
          spatial
        </td>
        <td
          rowspan="3"
        >
          üst
        </td>
        <td
          rowspan="1"
        >
          top
        </td>
        <td
          rowspan="3"
        >
          Gerekli olan her şey masanın üstünde.
        </td>
        <td
          rowspan="3"
        >
          Everything necessary is on the table
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          on top of
        </td>
        <td
          rowspan="1"
        >
          top
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          above
        </td>
        <td
          rowspan="1"
        >
          top
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="2"
        >
          arasında
        </td>
        <td
          rowspan="1"
        >
          between
        </td>
        <td
          rowspan="2"
        >
          posessive
        </td>
        <td
          rowspan="2"
        >
          spatial
        </td>
        <td
          rowspan="2"
        >
          ara
        </td>
        <td
          rowspan="1"
        >
          spaceinterval
        </td>
        <td
          rowspan="2"
        >
          Migros'la Ulus sineması arasında başka bina yok
        </td>
        <td
          rowspan="2"
        >
          There is no other building between Migros and the Ulus cinema
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          among
        </td>
        <td
          rowspan="1"
        >
          spaceinterval
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`works: 
<DocumentFragment>
  <table
    class="border-collapse table-auto"
  >
    <thead>
      <tr
        class="border-b border-gray-400 text-left"
      >
        <th>
          Word
        </th>
        <th>
          Definition
        </th>
        <th>
          Kind
        </th>
        <th>
          Group
        </th>
        <th
          colspan="2"
        >
          Root
        </th>
        <th
          colspan="2"
        >
          Examples
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          gibi
        </td>
        <td
          rowspan="1"
        >
          like
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe gibi
        </td>
        <td
          rowspan="1"
        >
          like Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          için
        </td>
        <td
          rowspan="1"
        >
          for
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe için
        </td>
        <td
          rowspan="1"
        >
          for Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          ile
        </td>
        <td
          rowspan="1"
        >
          with
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe ile
        </td>
        <td
          rowspan="1"
        >
          with Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          kadar
        </td>
        <td
          rowspan="1"
        >
          as ... as
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe kadar ...
        </td>
        <td
          rowspan="1"
        >
          as ... as Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="2"
        >
          doğru
        </td>
        <td
          rowspan="2"
        >
          towards (time/place)
        </td>
        <td
          rowspan="2"
        >
          bare
        </td>
        <td
          rowspan="2"
        >
          dative
        </td>
        <td
          rowspan="2"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          üçe doğru
        </td>
        <td
          rowspan="1"
        >
          towards 3 o'clock
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          eve doğru
        </td>
        <td
          rowspan="1"
        >
          towards home
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="3"
        >
          göre
        </td>
        <td
          rowspan="1"
        >
          according to
        </td>
        <td
          rowspan="3"
        >
          bare
        </td>
        <td
          rowspan="3"
        >
          dative
        </td>
        <td
          rowspan="3"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          sana göre
        </td>
        <td
          rowspan="1"
        >
          according to you
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          compared with
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          İngeltere'ye göre
        </td>
        <td
          rowspan="1"
        >
          compared with England
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          right for
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          bize göre
        </td>
        <td
          rowspan="1"
        >
          right for us
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="2"
        >
          kadar
        </td>
        <td
          rowspan="1"
        >
          until, by (time)
        </td>
        <td
          rowspan="2"
        >
          bare
        </td>
        <td
          rowspan="2"
        >
          dative
        </td>
        <td
          rowspan="2"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          sabah kadar
        </td>
        <td
          rowspan="1"
        >
          until morning
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          as far as (place)
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ankara'ya kadar
        </td>
        <td
          rowspan="1"
        >
          as far as Ankara
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="3"
        >
          üstünde
        </td>
        <td
          rowspan="1"
        >
          on
        </td>
        <td
          rowspan="3"
        >
          posessive
        </td>
        <td
          rowspan="3"
        >
          spatial
        </td>
        <td
          rowspan="3"
        >
          üst
        </td>
        <td
          rowspan="1"
        >
          top
        </td>
        <td
          rowspan="3"
        >
          Gerekli olan her şey masanın üstünde.
        </td>
        <td
          rowspan="3"
        >
          Everything necessary is on the table
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          on top of
        </td>
        <td
          rowspan="1"
        >
          top
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          above
        </td>
        <td
          rowspan="1"
        >
          top
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="2"
        >
          arasında
        </td>
        <td
          rowspan="1"
        >
          between
        </td>
        <td
          rowspan="2"
        >
          posessive
        </td>
        <td
          rowspan="2"
        >
          spatial
        </td>
        <td
          rowspan="2"
        >
          ara
        </td>
        <td
          rowspan="1"
        >
          spaceinterval
        </td>
        <td
          rowspan="2"
        >
          Migros'la Ulus sineması arasında başka bina yok
        </td>
        <td
          rowspan="2"
        >
          There is no other building between Migros and the Ulus cinema
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          among
        </td>
        <td
          rowspan="1"
        >
          spaceinterval
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
 1`] = `
<DocumentFragment>
  <table
    class="border-collapse table-auto"
  >
    <thead>
      <tr
        class="border-b border-gray-400 text-left"
      >
        <th>
          Word
        </th>
        <th>
          Definition
        </th>
        <th>
          Kind
        </th>
        <th>
          Group
        </th>
        <th
          colspan="2"
        >
          Root
        </th>
        <th
          colspan="2"
        >
          Examples
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          gibi
        </td>
        <td
          rowspan="1"
        >
          like
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe gibi
        </td>
        <td
          rowspan="1"
        >
          like Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          için
        </td>
        <td
          rowspan="1"
        >
          for
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe için
        </td>
        <td
          rowspan="1"
        >
          for Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          ile
        </td>
        <td
          rowspan="1"
        >
          with
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe ile
        </td>
        <td
          rowspan="1"
        >
          with Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          kadar
        </td>
        <td
          rowspan="1"
        >
          as ... as
        </td>
        <td
          rowspan="1"
        >
          bare
        </td>
        <td
          rowspan="1"
        >
          none/genitive
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ayşe kadar ...
        </td>
        <td
          rowspan="1"
        >
          as ... as Ayşe
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="2"
        >
          doğru
        </td>
        <td
          rowspan="2"
        >
          towards (time/place)
        </td>
        <td
          rowspan="2"
        >
          bare
        </td>
        <td
          rowspan="2"
        >
          dative
        </td>
        <td
          rowspan="2"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          üçe doğru
        </td>
        <td
          rowspan="1"
        >
          towards 3 o'clock
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          eve doğru
        </td>
        <td
          rowspan="1"
        >
          towards home
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="3"
        >
          göre
        </td>
        <td
          rowspan="1"
        >
          according to
        </td>
        <td
          rowspan="3"
        >
          bare
        </td>
        <td
          rowspan="3"
        >
          dative
        </td>
        <td
          rowspan="3"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          sana göre
        </td>
        <td
          rowspan="1"
        >
          according to you
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          compared with
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          İngeltere'ye göre
        </td>
        <td
          rowspan="1"
        >
          compared with England
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          right for
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          bize göre
        </td>
        <td
          rowspan="1"
        >
          right for us
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="2"
        >
          kadar
        </td>
        <td
          rowspan="1"
        >
          until, by (time)
        </td>
        <td
          rowspan="2"
        >
          bare
        </td>
        <td
          rowspan="2"
        >
          dative
        </td>
        <td
          rowspan="2"
        />
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          sabah kadar
        </td>
        <td
          rowspan="1"
        >
          until morning
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          as far as (place)
        </td>
        <td
          rowspan="1"
        />
        <td
          rowspan="1"
        >
          Ankara'ya kadar
        </td>
        <td
          rowspan="1"
        >
          as far as Ankara
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="3"
        >
          üstünde
        </td>
        <td
          rowspan="1"
        >
          on
        </td>
        <td
          rowspan="3"
        >
          posessive
        </td>
        <td
          rowspan="3"
        >
          spatial
        </td>
        <td
          rowspan="3"
        >
          üst
        </td>
        <td
          rowspan="1"
        >
          top
        </td>
        <td
          rowspan="3"
        >
          Gerekli olan her şey masanın üstünde.
        </td>
        <td
          rowspan="3"
        >
          Everything necessary is on the table
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          on top of
        </td>
        <td
          rowspan="1"
        >
          top
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          above
        </td>
        <td
          rowspan="1"
        >
          top
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="2"
        >
          arasında
        </td>
        <td
          rowspan="1"
        >
          between
        </td>
        <td
          rowspan="2"
        >
          posessive
        </td>
        <td
          rowspan="2"
        >
          spatial
        </td>
        <td
          rowspan="2"
        >
          ara
        </td>
        <td
          rowspan="1"
        >
          spaceinterval
        </td>
        <td
          rowspan="2"
        >
          Migros'la Ulus sineması arasında başka bina yok
        </td>
        <td
          rowspan="2"
        >
          There is no other building between Migros and the Ulus cinema
        </td>
      </tr>
      <tr
        class="border-b border-b-gray-200"
      >
        <td
          rowspan="1"
        >
          among
        </td>
        <td
          rowspan="1"
        >
          spaceinterval
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
